#327 Akka Streams Circuit Breaker

anilgursel opened the pull request:
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

Introduces Akka Streams BidiFlow to add Circuit Breaker functionality.  Also addresses the following issues:

- https://github.com/paypal/squbs/issues/374
- https://github.com/paypal/squbs/issues/359
- https://github.com/paypal/squbs/issues/361

Follow up issues:

- https://github.com/paypal/squbs/issues/373
- https://github.com/paypal/squbs/issues/372

-

akara commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/impl/AtomicCircuitBreakerState.scala:
```diff
@@ -0,0 +1,432 @@
+    * Holds reference to current resetTimeout of CircuitBreaker - *access only via helper methods*
+    */
+  @volatile
+  private[this] var _currentResetTimeoutDoNotCallMeDirectly: FiniteDuration = resetTimeout
```
What did you want to signify with the underbar convention?

anilgursel replied:
This is just a copy from the Akka Circuit Breaker.  Not sure what's signified.  I can drop the underbar?

akara commented on squbs-ext/src/test/java/org/squbs/circuitbreaker/CircuitBreakerBidiFlowTest.java:
```diff
@@ -0,0 +1,291 @@
+
+    @Test
+    public void testIncrementFailureCountBasedOnTheProvidedFunction() {
+        final ActorRef delayActor = system.actorOf(Props.create(org.squbs.streams.DelayActor.class));
```
Might want to address this one from Codacy.

anilgursel replied:
👍

akara commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerBidi.scala:
```diff
@@ -0,0 +1,307 @@
+import akka.http.org.squbs.util.JavaConverters.toJava
+import akka.japi.Pair
+import akka.stream._
+import akka.stream.scaladsl.BidiFlow
```
Can be confusing, but use the IDE's optimize import option to clean this up.

anilgursel replied:
👍

akara commented on docs/circuitbreaker.md:
```diff
@@ -0,0 +1,291 @@
+
+#### Creating CircuitBreakerState from configuration
+
+While `AtomicCircuitBreakerState` has programmatic API for configuration, it also allows the configuration to be provided in configuration:
```
Sounds odd: "it also allows the configuration to be provided in configuration." Should it be "it also allows the configuration to be provided through the system configuration."?

anilgursel replied:
👍

Also, chaning the title to "Creating CircuitBreakerState from system configuration".

akara commented on docs/circuitbreaker.md:
```diff
@@ -0,0 +1,291 @@
+
+#### Potential effects of Circuit Open state on throughput
+
+If the upstream does not control the throughput, then the throughput of the stream might temporarily increase once the circuit is `Open`:  The downstream demand will be addressed with short circuit/fallback messages, which might (or might not) take less time than it takes the joined `Flow` to process an element.  To eliminate this problem, a throttle can be applied specifically for circuit breake ...
```
... a throttle can be applied specifically for the circuit breaker open message: `(context, Failure(CircuitBreakerOpenException()))`.

anilgursel replied:
Modified it as "a throttle can be applied specifically for circuit breaker `Open` messages.".  Not using the code, because it might have to be done for fallback responses as well, and with that the example gets complicated.

anilgursel commented on docs/circuitbreaker.md:
```diff
@@ -0,0 +1,291 @@
+Circuit is `Closed`:
+
+   * If an output `msg` is provided by the wrapped flow within the timeout, then `Success(msg)` is passed down.
+   * Otherwise, a `Failure(FlowTimeoutException())` is pushed to downstream.
```
Updating it as `(Success(msg), context)` and `(Failure(FlowTimeoutException()), context)`.

anilgursel commented on docs/circuitbreaker.md:
```diff
@@ -0,0 +1,291 @@
+Circuit is `Open`:
+
+   * The result of `fallback` function is pushed to downstream if one provided
+   * Otherwise, a `Failure(CircuitBreakerOpenException())` is pushed to downstream 	
```
👍

akara commented on squbs-ext/src/test/scala/org/squbs/circuitbreaker/CircuitBreakerBidiFlowSpec.scala:
```diff
@@ -0,0 +1,375 @@
+      import context.dispatcher
+      context.system.scheduler.scheduleOnce(delay(element), sender(), element)
+    case element: (String, Long) =>
+      import context.dispatcher
```
Just import `context.dispatcher` once inside this class. Makes code easier to read.

anilgursel replied:
👍

akara commented on squbs-ext/src/test/scala/org/squbs/circuitbreaker/CircuitBreakerBidiFlowSpec.scala:
```diff
@@ -0,0 +1,375 @@
+    case element: String =>
+      import context.dispatcher
+      context.system.scheduler.scheduleOnce(delay(element), sender(), element)
+    case element: (String, Long) =>
```
```scala
case (element: String, _) => context.system.scheduler.scheduleOnce(delay(element), sender(), element)
```

You can also dismiss the third case match altogether.

anilgursel replied:
👍

anilgursel replied:
Actually, it needs to be `case element: (String, _) => context.system.scheduler.scheduleOnce(delay(element._1), sender(), element)`.

akara commented on squbs-ext/src/test/scala/org/squbs/circuitbreaker/CircuitBreakerBidiFlowSpec.scala:
```diff
@@ -0,0 +1,375 @@
+    val failFastFailure = Failure(CircuitBreakerOpenException())
+    val expected = (Success("a"), 1) :: (flowFailure, 2) :: (flowFailure, 3) :: (failFastFailure, 4) :: Nil
+    whenReady(result) { r =>
+      r should contain theSameElementsInOrderAs(expected)
```
```scala
r should contain theSameElementsInOrderAs expected
```

anilgursel replied:
👍

akara commented on squbs-ext/src/test/scala/org/squbs/circuitbreaker/CircuitBreakerBidiFlowSpec.scala:
```diff
@@ -0,0 +1,375 @@
+    expectMsg(Closed)
+  }
+
+  it should "increment failure count based on the provided function" in {
```
Hmm, the description may not be accurate. Are we really checking the failure count in this test?

anilgursel replied:
how about `it should "decide on failures based on the provided function" in {`?

anilgursel commented on squbs-ext/src/main/java/org/squbs/circuitbreaker/impl/AbstractAtomicCircuitBreakerLogic.java:
```diff
@@ -0,0 +1,36 @@
+            resetTimeoutOffset = Unsafe.instance.objectFieldOffset(AtomicCircuitBreakerState.class.getDeclaredField("_currentResetTimeoutDoNotCallMeDirectly"));
+        } catch(Throwable t){
+            throw new ExceptionInInitializerError(t);
+        }
```
Actually, this code is mostly taken from Akka Pattern.  Please see https://github.com/akka/akka/blob/master/akka-actor/src/main/java/akka/pattern/AbstractCircuitBreaker.java.  We can try different alternatives, but in the first cut, I did not want to invest the time on that and just rely on existing logic.

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerState.scala:
```diff
@@ -0,0 +1,233 @@
+  /**
+    * Check if circuit should be short circuited.
+    */
+  final def shortCircuited(): Boolean = {
```
it actually updates metrics.  Doesn't it count?

anilgursel replied:
> This applies to multiple defs in this class.

What other `def`?

akara replied:
Hmm, the query into `shortCircuited` causes a `meter` to increase? This is on the state object, highly shared. Is it the right place to increment the `ShortCircuitCount`? I'm just thinking out loud whether the method name `shortCircuited` is right. It should imply you're asking for a decision and record that decision in the metrics instead of a casual name. Some name (for lack of a better name) like `checkAndRecordShortCircuit()` would be more meaningful in this case.

anilgursel replied:
Metric wise, this is the right place if we would like to meter how many times it preferred the short route.

Naming wise, if we rename it as `checkAndRecordShortCircuit`, then we need to do it for  `success` and `failure`  methods as well, which kinda makes the API less clean.  And in case we decide to drop metrics or provide it in a different way in the future (just assuming), then the API would need to change again.

anilgursel replied:
Copying my comment from Gitter https://gitter.im/paypal/squbs?at=589b36f96b2d8dd55219d3c5:

> How about the following method names?
>
> - markSuccess
> - markFailure
> - checkAndMarkIfShortCircuit

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerState.scala:
```diff
@@ -0,0 +1,233 @@
+    * Resets breaker to `Closed` state.  This is valid from `HalfOpen` state only.
+    *
+    */
+  protected final def resetBreaker(): Unit = transition(HalfOpen, Closed)
```
If it is already in the state that it wants to transition to, then the impl returns `false`, and `transition` just skips publishing state changed event.

```scala
protected final def transition(fromState: State, toState: State): Unit =
    if(transitionImpl(fromState, toState)) eventBus.publish(CircuitBreakerEvent(toState, toState))
```

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerBidi.scala:
```diff
@@ -0,0 +1,305 @@
+  private val toWrapped = Outlet[(In, Context)]("CircuitBreakerBidi.toWrapped")
+  private val out = Outlet[(Try[Out], Context)]("CircuitBreakerBidi.out")
+  val shape = BidiShape(in, toWrapped, fromWrapped, out)
+  private var upstreamFinished = false
```
Thanks for catching!  The following three should actually move inside `createLogic`:

```scala
private var upstreamFinished = false
    private val readyToPush = mutable.Queue[(Try[Out], Context)]()

    private def onPushFromWrapped(elem: (Try[Out], Context), isOutAvailable: Boolean): Option[(Try[Out], Context)] = {
      readyToPush.enqueue(elem)
      if(isOutAvailable) Some(readyToPush.dequeue())
      else None
    }
```

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerBidi.scala:
```diff
@@ -0,0 +1,305 @@
+  private val isFailure = failureDecider.getOrElse((e: (Try[Out], Context)) => e._1.isFailure)
+
+  private def onPushFromWrapped(elem: (Try[Out], Context), isOutAvailable: Boolean): Option[(Try[Out], Context)] = {
+    readyToPush.enqueue(elem)
```
👍

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerBidi.scala:
```diff
@@ -0,0 +1,305 @@
+
+    setHandler(fromWrapped, new InHandler {
+      override def onPush(): Unit = {
+        val elemWithcontext = grab(fromWrapped)
```
IntelliJ also warns!  Somehow overlooked it.

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/CircuitBreakerBidi.scala:
```diff
@@ -0,0 +1,305 @@
+
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit =
+        if(!upstreamFinished || !readyToPush.isEmpty) readyToPush.dequeueFirst((_: (Try[Out], Context)) => true) match {
```
👍

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/impl/AtomicCircuitBreakerState.scala:
```diff
@@ -0,0 +1,432 @@
+    */
+  @inline
+  private[this] def swapAtomicState(oldState: AtomicState, newState: AtomicState): Boolean =
+  Unsafe.instance.compareAndSwapObject(this, AbstractAtomicCircuitBreakerLogic.stateOffset, oldState, newState)
```
I think in the second round, we should try to avoid `Unsafe`.  We can re-visit this at that point?

anilgursel commented on squbs-ext/src/main/scala/org/squbs/circuitbreaker/impl/AtomicCircuitBreakerState.scala:
```diff
@@ -0,0 +1,432 @@
+  /**
+    * Check if circuit should be short circuited.
+    */
+  def isShortCircuited(): Boolean = currentAtomicState.isShortCircuited
```
👍

anilgursel commented on squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala:
```diff
@@ -23,153 +23,151 @@ import akka.http.org.squbs.util.JavaConverters._
-        }
+  setHandler(fromWrapped, new InHandler {
+    override def onPush(): Unit = {
+      onPushFromWrapped(grab(fromWrapped), isAvailable(out)) map { elem =>
```
👍

anilgursel commented:
Pushed a new commit that addresses the review comments.  I dropped some questions for some of the comments though, let's discuss those.  

Having some intermittent test failures, looking into those.
